File: ofdrw-java/ofdrw-core/src/main/java/org/ofdrw/core/basicType/ST_Loc.java
package org.ofdrw.core.basicType;


import org.dom4j.Element;

import java.util.ArrayList;

/**
 * 包结构内文件的路径，“.”表示当前路径，“..”表示符路径
 * <p>
 * 约定：
 * 1. “、”代表根节点；
 * 2. 未显示指定代表当前路径；
 * 3. 路径区分大小写
 * <p>
 * 示例：
 * <code>
 * /Pages/P1/Content.xml
 * ./Res/Book1.jpg
 * ../Pages/P1/Res.xml
 * Pages/P1/Rcs.xml
 * </code>
 * ————《GB/T 33190-2016》 表 2 基本数据类型
 *
 * @author 权观宇
 * @since 2019-09-28 10:40:33
 */
public class ST_Loc extends STBase {
    /**
     * 路径
     */
    private String loc;

    public ST_Loc(String path) {
        this.loc = path;
    }

    /**
     * 获取路径实例
     *
     * @param loc 路径字符串
     * @return 实例或null
     */
    public static ST_Loc getInstance(String loc) {
        if (loc == null || loc.trim().length() == 0) {
            return null;
        }
        return new ST_Loc(loc.trim());
    }

    /**
     * 从实例中获取 路径
     *
     * @param e 元素
     * @return 路径对象
     */
    public static ST_Loc getInstance(Element e) {
        if (e == null) {
            return null;
        }
        return getInstance(e.getTextTrim());
    }

    public String getLoc() {
        return loc;
    }

    public ST_Loc setLoc(String loc) {
        this.loc = loc;
        return this;
    }

    /**
     * 路径分割
     * <p>
     * 通过该方法获取的路径中可能含有 "" 空字符串，
     * 如果需要获取路径的各个部分，请使用 {@link #parts()}
     *
     * @return 各个子路径
     */
    public String[] split() {
        if (loc == null) {
            return new String[0];
        }
        return this.loc.split("/");
    }

    /**
     * 获取路径的各个部分
     *
     * @return 路径的各个部分
     */
    public String[] parts() {
        if (loc == null) {
            return new String[0];
        }
        String[] split = this.loc.split("/");
        // 移除split中为空的元素
        ArrayList<String> paths = new ArrayList<>();
        for (String s : split) {
            if (s != null && s.trim().length() > 0) {
                paths.add(s);
            }
        }
        return paths.toArray(new String[0]);
    }

    /**
     * 获取父母路径
     *
     * @return 父母路径字符串
     */
    public String parent() {
        int indexOf = loc.lastIndexOf('/');
        if (indexOf == -1) {
            return "";
        }
        return loc.substring(0, indexOf);
    }

    /**
     * 获取父母路径
     *
     * @return 父母路径
     */
    public ST_Loc parentLoc() {
        return new ST_Loc(this.parent());
    }

    /**
     * 获取路径的文件名称
     *
     * @return 文件名称
     */
    public String getFileName() {
        int indexOf = loc.lastIndexOf('/');
        if (indexOf == -1) {
            return loc;
        } else if (indexOf == loc.length() - 1) {
            return "";
        }
        return loc.substring(indexOf + 1);
    }

    /**
     * 路径拼接
     *
     * @param p2 路径对象
     * @return 拼接后路径
     */
    public ST_Loc cat(String p2) {
        if (p2 == null) {
            return this;
        }

        String part1 = this.loc;
        if (part1.endsWith("/")) {
            part1 = part1.substring(0, part1.length() - 1);
        }
        if (p2.startsWith("/")) {
            p2 = p2.substring(1);
        }
        return new ST_Loc(part1 + "/" + p2);
    }

    /**
     * 路径拼接
     *
     * @param loc 路径对象
     * @return 拼接后路径
     */
    public ST_Loc cat(ST_Loc loc) {
        if (loc == null) {
            return this;
        }
        return cat(loc.getLoc());
    }

    /**
     * 是否以指定字符结尾
     *
     * @param suffix 指定字符
     * @return true 指定字符结尾；false 不以指定字符结尾
     */
    public boolean endWith(String suffix) {
        return loc.endsWith(suffix);
    }

    /**
     * 是否以指定字符开头
     *
     * @param prefix 前缀
     * @return true 指定字符开头；false 不以指定字符开头
     */
    public boolean startWith(String prefix) {
        return loc.startsWith(prefix);
    }

    /**
     * 是否为根路径（以"/"开头）
     *
     * @return true - 根路径
     */
    public boolean isRootPath() {
        return loc.startsWith("/");
    }

    /**
     * 是否为空或空串
     *
     * @return true - 空；false - 非空
     */
    public boolean isEmpty() {
        return loc == null || loc.trim().length() == 0;
    }

    @Override
    public String toString() {
        return loc;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof String) {
            return loc.equals(obj);
        }
        if (obj instanceof ST_Loc) {
            return loc.equals(((ST_Loc) obj).loc);
        }
        return loc.equals(obj.toString());
    }

    /**
     * 比较两个路径是否相等
     *
     * @param a 比较对象a
     * @param b 比较对象b
     * @return true - 相等，false - 不相等
     */
    public static boolean equal(Object a, Object b) {
        if (a == null || b == null) {
            return false;
        }
        if (a instanceof ST_Loc && b instanceof ST_Loc) {
            return ((ST_Loc) a).loc.equals(((ST_Loc) b).loc);
        }
        String aStr = a.toString();
        String bStr = b.toString();
        return aStr.equals(bStr);
    }
}

File: OfdrwNet.Core/BasicType/StLoc.cs
namespace OfdrwNet.Core.BasicType;

/// <summary>
/// ST_Loc 位置类型
/// 对应 Java 版本的 org.ofdrw.core.basicType.ST_Loc
/// 用于表示文件路径或资源位置
/// </summary>
public class StLoc
{
    /// <summary>
    /// 位置路径
    /// </summary>
    public string Path { get; set; }

    /// <summary>
    /// 初始化位置
    /// </summary>
    /// <param name="path">路径字符串</param>
    public StLoc(string path)
    {
        Path = path ?? throw new ArgumentNullException(nameof(path));
    }

    /// <summary>
    /// 连接路径
    /// </summary>
    /// <param name="subPath">子路径</param>
    /// <returns>新的位置实例</returns>
    public StLoc Cat(string subPath)
    {
        if (string.IsNullOrEmpty(subPath))
            return this;
        
        string separator = Path.EndsWith("/") || subPath.StartsWith("/") ? "" : "/";
        return new StLoc($"{Path}{separator}{subPath}");
    }

    /// <summary>
    /// 获取父级路径
    /// </summary>
    /// <returns>父级位置，如果没有父级则返回当前位置</returns>
    public StLoc Parent()
    {
        int lastSlash = Path.LastIndexOf('/');
        if (lastSlash > 0)
        {
            return new StLoc(Path.Substring(0, lastSlash));
        }
        return new StLoc("/");
    }

    /// <summary>
    /// 获取文件名
    /// </summary>
    /// <returns>文件名</returns>
    public string GetFileName()
    {
        int lastSlash = Path.LastIndexOf('/');
        if (lastSlash >= 0 && lastSlash < Path.Length - 1)
        {
            return Path.Substring(lastSlash + 1);
        }
        return Path;
    }

    /// <summary>
    /// 转换为字符串
    /// </summary>
    /// <returns>路径字符串</returns>
    public override string ToString()
    {
        return Path;
    }

    /// <summary>
    /// 隐式转换为字符串
    /// </summary>
    /// <param name="stLoc">位置实例</param>
    /// <returns>路径字符串</returns>
    public static implicit operator string(StLoc stLoc)
    {
        return stLoc.Path;
    }

    /// <summary>
    /// 隐式转换从字符串
    /// </summary>
    /// <param name="path">路径字符串</param>
    /// <returns>位置实例</returns>
    public static implicit operator StLoc(string path)
    {
        return new StLoc(path);
    }

    /// <summary>
    /// 从字符串解析位置
    /// </summary>
    /// <param name="path">路径字符串</param>
    /// <returns>位置实例</returns>
    public static StLoc Parse(string path)
    {
        if (string.IsNullOrEmpty(path))
            throw new ArgumentException("路径不能为空", nameof(path));
        return new StLoc(path);
    }

    /// <summary>
    /// 相等比较
    /// </summary>
    /// <param name="obj">比较对象</param>
    /// <returns>是否相等</returns>
    public override bool Equals(object? obj)
    {
        if (obj is StLoc other)
            return Path == other.Path;
        if (obj is string stringValue)
            return Path == stringValue;
        return false;
    }

    /// <summary>
    /// 获取哈希码
    /// </summary>
    /// <returns>哈希码</returns>
    public override int GetHashCode()
    {
        return Path.GetHashCode();
    }

    /// <summary>
    /// 创建 StLoc 的实例（兼容现有代码习惯）
    /// </summary>
    public static StLoc GetInstance(string path)
    {
        return new StLoc(path);
    }
    
    /// <summary>
    /// 判断是否为根路径（以 / 开头）
    /// </summary>
    public bool IsRootPath()
    {
        return !string.IsNullOrEmpty(Path) && Path.StartsWith("/");
    }
    
    /// <summary>
    /// 比较字符串路径与 StLoc 是否相同（兼容原 Java 风格 Equal）
    /// </summary>
    public static bool Equal(string? a, StLoc? b)
    {
        if (a == null && b == null) return true;
        if (a == null || b == null) return false;
        return a == b.Path;
    }
    
    /// <summary>
    /// 连接另一个 StLoc
    /// </summary>
    public StLoc Cat(StLoc? other)
    {
        if (other == null) return this;
        return Cat(other.Path);
    }
}